pyfr/plugins/tavg.py

```diff
@@ -358,3 +363,247 @@ def __call__(self, intg):
+            smap = [i for n, i in sorted(zip(snames, sidx))]
+
+            if i == 0:
+                self.stats, self.cfg = stats, cfg
+
+                prec = cfg.get('backend', 'precision')
+                self.dtype = np.float32() if prec == 'single' else np.float64()
+
+                self.fields = fields
+                self.cfgsect = cfgsect
+                self.region = cfg.get(cfgsect, 'region')
+                self.uuid = file['mesh_uuid']
+
+                self.idxs = [k for k in file if re.search(r'_idxs_p\d+$', k)]
+                self.dkeys = [k for k in file if
+                              re.match(r'tavg_[^\W_]+_p\d+$', k)]
+
+                self.mapping = [(amap, smap)]
+                fnames = [x.removeprefix('fun-avg-') for x in fields if
+                          x.startswith('fun-avg-')]
+                self.anames = sorted(anames)
+                self.fnames = sorted(fnames)
+
+                # Initialise std max and avg registers
+                if self.std_all:
+                    self.std_max = np.zeros(len(anames))
+                    self.std_avg = np.zeros(len(anames))
+                    self.fstd_max = np.zeros(len(fnames))
+                    self.fstd_avg = np.zeros(len(fnames))
+
+                # Build fun avg expressions
+                c = cfg.items_as('constants', float)
+                self.fexprs = [cfg.getexpr(cfgsect, f'fun-avg-{k}', subs=c)
+                               for k in fnames]
+            else:
+                self.mapping.append((amap, smap))
+
+            # Check for compatibility of files
+            if self.uuid != file['mesh_uuid']:
+                raise RuntimeError('Average files computed on different meshes')
+            if self.region != cfg.get(cfgsect, 'region'):
+                raise RuntimeError('Average files computed on different regions')
+
+            for k in cfg.items(cfgsect, prefix='avg-'):
+                if cfg.get(cfgsect, k) != self.cfg.get(self.cfgsect, k):
+                    raise RuntimeError('Different average field definitions')
+
+    def _merge_data(self, outf):
+        file0, stats, cfg, dt = self.files[0]
+        amap0, smap0 = self.mapping[0]
+
+        for key in self.dkeys:
+            # Initialise accumulators
+            data = file0[key]
+            avg_acc = np.zeros_like(data[:, amap0])
+            if self.std_all:
+                var_acc = np.zeros_like(data[:, smap0])
+
+            # Perform the rest of the accumulations
+            for files, mapping in zip(self.files, self.mapping):
+                file, stats, cfg, dt = files
+                amap, smap = mapping
+                data = file[key]
+
+                avg_acc += dt*data[:, amap]
+                if self.std_all:
+                    var_acc += dt*(data[:, smap]**2 + data[:, amap]**2)
+
+            # Shift variance and get standard deviation
+            if self.std_all:
+                var_acc = np.sqrt(var_acc - avg_acc**2)
```

Comment:
Can you check if this approach is numerically stable (at least on paper)? For some reason I think it is the 'bad' way to do it.